flocker/control/test/test_persistence.py

```diff
@@ -249,3 +338,199 @@ def test_complex_keys(self):
+        # Compare the v1 --> v3 upgrade to the direct result of the
+        # v2 --> v3 upgrade on the v2 config, Both should be identical
+        # and valid v3 configs.
+        self.assertEqual(result, StubMigration.upgrade_from_v2(v2_config))
+
+
+UUIDS = st.basic(generate=lambda r, _: UUID(int=r.getrandbits(128)))
+
+DATASETS = st.builds(Dataset, dataset_id=UUIDS, maximum_size=st.integers())
+
+# Constrain primary to be True so that we don't get invariant errors from Node
+# due to having two differing manifestations of the same dataset id.
+MANIFESTATIONS = st.builds(
+    Manifestation, primary=st.just(True), dataset=DATASETS)
+IMAGES = st.builds(DockerImage, tag=st.text(), repository=st.text())
+NONE_OR_INT = st.one_of(
+    st.none(),
+    st.integers()
+)
+ST_PORTS = st.integers(min_value=1, max_value=65535)
+PORTS = st.builds(
+    Port,
+    internal_port=ST_PORTS,
+    external_port=ST_PORTS
+)
+LINKS = st.builds(
+    Link,
+    local_port=ST_PORTS,
+    remote_port=ST_PORTS,
+    alias=st.text(alphabet=string.letters, min_size=1)
+)
+FILEPATHS = st.text(alphabet=string.printable).map(FilePath)
+VOLUMES = st.builds(
+    AttachedVolume, manifestation=MANIFESTATIONS, mountpoint=FILEPATHS)
+APPLICATIONS = st.builds(
+    Application, name=st.text(), image=IMAGES,
+    # A MemoryError will likely occur without the max_size limits on
+    # Ports and Links. The max_size value that will trigger memory errors
+    # will vary system to system. 10 is a reasonable test range for realistic
+    # container usage that should also not run out of memory on most modern
+    # systems.
+    ports=st.sets(PORTS, max_size=10),
+    links=st.sets(LINKS, max_size=10),
+    volume=st.none() | VOLUMES,
+    environment=st.dictionaries(keys=st.text(), values=st.text()),
+    memory_limit=NONE_OR_INT,
+    cpu_shares=NONE_OR_INT,
+    running=st.booleans()
+)
+
+
+def _build_node(applications):
+    # All the manifestations in `applications`.
+    app_manifestations = set(
+        app.volume.manifestation for app in applications if app.volume
+    )
+    # A set that contains all of those, plus an arbitrary set of
+    # manifestations.
+    dataset_ids = frozenset(
+        app.volume.manifestation.dataset_id
+        for app in applications if app.volume
+    )
+    manifestations = (
+        st.sets(MANIFESTATIONS.filter(
+            lambda m: m.dataset_id not in dataset_ids))
+        .map(pset)
+        .map(lambda ms: ms.union(app_manifestations))
+        .map(lambda ms: dict((m.dataset.dataset_id, m) for m in ms)))
+    return st.builds(
+        Node, uuid=UUIDS,
+        applications=st.just(applications),
+        manifestations=manifestations)
+
+
+NODES = st.sets(APPLICATIONS).map(
+    lambda apps: pset(dict((
+        app.volume.manifestation.dataset_id, app)
+        for app in apps if app.volume).values())
+).flatmap(_build_node)
+
+
+DEPLOYMENTS = st.builds(
+    Deployment, nodes=st.sets(NODES, min_size=0, max_size=3)
```

Comment:
1. You never need `min_size=0` on a set. No reasonable reader is going to assume the default is anything but. Please drop it.
2. Please either delete the maximum size constraint, or give a concrete reason in the comments for constraining `max_size` to 3, explaining why 3 was chosen.

(I'm not 100% sure, but think I made these comment in a previous pass)

The more general point is that code should communicate intent. No one is going to be able to look at `max_size=3` and guess why we chose 3, or what we were trying achieve by doing so.